odl/operator/operator.py

```diff
@@ -93,42 +77,247 @@ def _default_apply(self, x, out, *args, **kwargs):
     -------
     `None`
     """
-    out.assign(self._call(x, *args, **kwargs))
+    out.assign(op._call_out_of_place(x, **kwargs))
 
 
-class _OperatorMeta(type):
+def _signature_from_spec(func):
+    """Return the signature of a python function as a string.
 
-    """Metaclass used by `Operator` to ensure correct methods.
+    Parameters
+    ----------
+    func : `function`
+        Function whose signature to compile
 
-    If either ``_apply()`` or ``_call()`` does not exist in the class
-    to be created, this metaclass attempts to add a default
-    implmentation.
-    This only works if the `Operator.range` is a
-    `LinearSpace`.
+    Returns
+    -------
+    sig : `str`
+        Signature of the function
     """
+    import sys
+    import inspect
 
-    def __new__(mcs, name, bases, attrs):
-        """Create a new instance."""
-        if '_call' in attrs and '_apply' in attrs:
-            pass
-        elif '_call' in attrs:
-            attrs['_apply'] = _default_apply
-        elif '_apply' in attrs:
-            attrs['_call'] = _default_call
+    py3 = (sys.version_info.major > 2)
+    if py3:
+        spec = inspect.getfullargspec(func)
+    else:
+        spec = inspect.getargspec(func)
+
+    posargs = spec.args
+    defaults = spec.defaults
+    kwargs = spec.varkw if py3 else spec.keywords
+    deflen = 0 if defaults is None else len(defaults)
+    nodeflen = 0 if posargs is None else len(posargs) - deflen
+    if nodeflen > 0:
+        argstr = ', '.join(posargs[:nodeflen])
+    else:
+        argstr = ''
+
+    if defaults:
+        if argstr:
+            argstr += ', '
+        argstr += ', '.join(['{}={}'.format(arg, dval)
+                             for arg, dval in zip(posargs[nodeflen:],
+                                                  defaults)])
+    if py3:
+        kw_only = spec.kwonlyargs
+        kw_only_defaults = spec.kwonlydefaults
+        if argstr:
+            argstr += ', '
+        if kw_only:
+            argstr += ', *, '
+            argstr += ', '.join('{}={}'.format(arg, dval)
+                                for arg, dval in zip(kw_only,
+                                                     kw_only_defaults))
+    if kwargs:
+        if argstr:
+            argstr += ', '
+        argstr += '**{}'.format(kwargs)
+
+    return '{}({})'.format(func.__name__, argstr)
+
+
+def _dispatch_call_args(cls=None, bound_call=None, unbound_call=None,
```

Comment:
This was apparently hacked quite quickly. It's just for pretty-printing the spec of a function, which for unknown reasons is not supported in the `inspect` module in Python 2. In the Py3 counterpart, there are much more functions, among them a pretty-printer.

Whatever, I'll write tests for this guy as well.